File: src/main/java/io/github/takenoha/towerdefense/runtime/CoreBlockKey.java
package io.github.takenoha.towerdefense.runtime;

import java.util.Objects;
import java.util.UUID;
import org.bukkit.block.Block;

/** Exact world/block coordinate used to identify a persisted core without chunk access. */
public record CoreBlockKey(UUID worldId, int x, int y, int z) {
    public CoreBlockKey {
        Objects.requireNonNull(worldId, "worldId");
    }

    public static CoreBlockKey from(Block block) {
        Objects.requireNonNull(block, "block");
        return new CoreBlockKey(
                block.getWorld().getUID(),
                block.getX(),
                block.getY(),
                block.getZ());
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/runtime/CoreBlockKey.kt
package io.github.takenoha.towerdefense.runtime

import java.util.Objects
import java.util.UUID
import kotlin.jvm.JvmRecord
import org.bukkit.block.Block

/** Exact world/block coordinate used to identify a persisted core without chunk access. */
@JvmRecord
data class CoreBlockKey(
    val worldId: UUID,
    val x: Int,
    val y: Int,
    val z: Int,
) {
    init {
        Objects.requireNonNull(worldId, "worldId")
    }

    companion object {
        @JvmStatic
        fun from(block: Block): CoreBlockKey {
            Objects.requireNonNull(block, "block")
            return CoreBlockKey(block.world.uid, block.x, block.y, block.z)
        }
    }
}
